File: app/backend/api/app/address_service.py
from models import db, Address
from typing import List, Optional, Dict, Any

class AddressService:
    """地址服务层 - 封装地址相关的业务逻辑"""
    
    @staticmethod
    def get_user_addresses(user_id: int) -> List[Dict[str, Any]]:
        """获取用户的所有地址"""
        addresses = Address.query.filter_by(user_id=user_id).order_by(
            Address.is_default.desc(), 
            Address.created_at.desc()
        ).all()
        
        address_list = []
        for addr in addresses:
            address_list.append({
                'id': addr.id,
                'receiver_name': addr.receiver_name,
                'phone': addr.phone,
                'province': addr.province,
                'city': addr.city,
                'district': addr.district,
                'detail_address': addr.detail_address,
                'is_default': addr.is_default,
                'full_address': f"{addr.province}{addr.city}{addr.district}{addr.detail_address}"
            })
        
        return address_list
    
    @staticmethod
    def get_default_address(user_id: int) -> Optional[Dict[str, Any]]:
        """获取用户的默认地址"""
        default_address = Address.query.filter_by(
            user_id=user_id, 
            is_default=True
        ).first()
        
        if not default_address:
            return None
            
        return {
            'id': default_address.id,
            'receiver_name': default_address.receiver_name,
            'phone': default_address.phone,
            'province': default_address.province,
            'city': default_address.city,
            'district': default_address.district,
            'detail_address': default_address.detail_address,
            'is_default': default_address.is_default,
            'full_address': f"{default_address.province}{default_address.city}{default_address.district}{default_address.detail_address}"
        }
    
    @staticmethod
    def get_address_by_id(address_id: int, user_id: int) -> Optional[Dict[str, Any]]:
        """根据ID获取地址"""
        address = Address.query.filter_by(id=address_id, user_id=user_id).first()
        
        if not address:
            return None
            
        return {
            'id': address.id,
            'receiver_name': address.receiver_name,
            'phone': address.phone,
            'province': address.province,
            'city': address.city,
            'district': address.district,
            'detail_address': address.detail_address,
            'is_default': address.is_default,
            'full_address': f"{address.province}{address.city}{address.district}{address.detail_address}"
        }
    
    @staticmethod
    def create_address(user_id: int, address_data: Dict[str, Any]) -> Dict[str, Any]:
        """创建新地址"""
        # 如果设置为默认地址，先取消其他默认地址
        if address_data.get('is_default'):
            Address.query.filter_by(user_id=user_id, is_default=True).update({'is_default': False})
        
        new_address = Address(
            user_id=user_id,
            receiver_name=address_data['receiver_name'],
            phone=address_data['phone'],
            province=address_data['province'],
            city=address_data['city'],
            district=address_data['district'],
            detail_address=address_data['detail_address'],
            is_default=address_data.get('is_default', False)
        )
        
        db.session.add(new_address)
        db.session.commit()
        
        return {
            'id': new_address.id,
            'receiver_name': new_address.receiver_name,
            'phone': new_address.phone,
            'province': new_address.province,
            'city': new_address.city,
            'district': new_address.district,
            'detail_address': new_address.detail_address,
            'is_default': new_address.is_default,
            'full_address': f"{new_address.province}{new_address.city}{new_address.district}{new_address.detail_address}"
        }
    
    @staticmethod
    def update_address(address_id: int, user_id: int, address_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """更新地址"""
        address = Address.query.filter_by(id=address_id, user_id=user_id).first()
        
        if not address:
            return None
        
        # 如果设置为默认地址，先取消其他默认地址
        if address_data.get('is_default'):
            Address.query.filter_by(user_id=user_id, is_default=True).update({'is_default': False})
        
        # 更新地址信息
        for key, value in address_data.items():
            if hasattr(address, key):
                setattr(address, key, value)
        
        db.session.commit()
        
        return {
            'id': address.id,
            'receiver_name': address.receiver_name,
            'phone': address.phone,
            'province': address.province,
            'city': address.city,
            'district': address.district,
            'detail_address': address.detail_address,
            'is_default': address.is_default,
            'full_address': f"{address.province}{address.city}{address.district}{address.detail_address}"
        }
    
    @staticmethod
    def delete_address(address_id: int, user_id: int) -> bool:
        """删除地址"""
        address = Address.query.filter_by(id=address_id, user_id=user_id).first()
        
        if not address:
            return False
        
        db.session.delete(address)
        db.session.commit()
        
        return True
    
    @staticmethod
    def set_default_address(address_id: int, user_id: int) -> bool:
        """设置默认地址"""
        # 先取消所有默认地址
        Address.query.filter_by(user_id=user_id, is_default=True).update({'is_default': False})
        
        # 设置新的默认地址
        address = Address.query.filter_by(id=address_id, user_id=user_id).first()
        if not address:
            return False
        
        address.is_default = True
        db.session.commit()
        
        return True 

File: app/front-app-uni/src/services/addressService.js
import request from '@/src/utils/request.js'
import { getUserId } from '@/src/utils/auth.js'

/**
 * Address Service - 地址服务
 * 遵循DDD原则，封装地址相关的业务逻辑
 * 适配 uni-app 环境
 */
class AddressService {
	/**
	 * 获取用户默认地址
	 * 优先从本地存储获取，如果没有则从后端API获取并缓存
	 * @param {number} userId - 用户ID
	 * @returns {Promise<Object|null>} 默认地址对象或null
	 */
	static async getDefaultAddress(userId) {
		try {
			// 1. 优先从本地存储获取选中的地址
			const uid = userId || getUserId()
			const cachedAddress = this.getSelectedAddress()
			if (cachedAddress) {
				return cachedAddress
			}

			// 2. 如果本地存储中没有地址，则从后端获取默认地址
			const response = await request.get('/api/app/address/default', {
				params: { user_id: uid }
			})
			
			if (response.data.code === 200 && response.data.data) {
				const defaultAddress = response.data.data
				// 3. 缓存默认地址到本地存储
				this.setSelectedAddress(defaultAddress)
				return defaultAddress
			} else {
				console.error('获取默认地址失败:', response.data.message)
				return null
			}
		} catch (error) {
			console.error('获取默认地址异常:', error)
			return null
		}
	}

	/**
	 * 设置选中的地址到本地存储
	 * @param {Object} address - 地址对象
	 */
	static setSelectedAddress(address) {
		try {
			const uid = getUserId()
			if (address && uid) {
				// 使用 uni-app 的存储API
				uni.setStorageSync(`selectedAddress:${uid}`, JSON.stringify(address))
			}
		} catch (error) {
			console.error('保存地址到本地存储失败:', error)
		}
	}

	/**
	 * 清除选中的地址本地存储
	 */
	static clearSelectedAddress() {
		try {
			const uid = getUserId()
			if (uid) {
				uni.removeStorageSync(`selectedAddress:${uid}`)
			}
		} catch (error) {
			console.error('清除地址本地存储失败:', error)
		}
	}

	/**
	 * 从本地存储获取选中的地址
	 * @returns {Object|null} 地址对象或null
	 */
	static getSelectedAddress() {
		try {
			const uid = getUserId()
			if (!uid) return null
			
			const cachedAddress = uni.getStorageSync(`selectedAddress:${uid}`)
			if (cachedAddress) {
				return JSON.parse(cachedAddress)
			}
		} catch (error) {
			console.error('获取地址本地存储失败:', error)
			// 如果解析失败，清除损坏的数据
			this.clearSelectedAddress()
		}
		return null
	}

	/**
	 * 获取用户所有地址
	 * @param {number} userId - 用户ID
	 * @returns {Promise<Array>} 地址列表
	 */
	static async getUserAddresses(userId) {
		try {
			const uid = userId || getUserId()
			const response = await request.get('/api/app/address/', {
				params: { user_id: uid }
			})
			
			if (response.data.code === 200) {
				return response.data.data || []
			} else {
				console.error('获取地址列表失败:', response.data.message)
				return []
			}
		} catch (error) {
			console.error('获取地址列表异常:', error)
			return []
		}
	}

	/**
	 * 添加新地址
	 * @param {Object} addressData - 地址数据
	 * @returns {Promise<Object|null>} 添加结果
	 */
	static async addAddress(addressData) {
		try {
			const uid = getUserId()
			if (!uid) {
				throw new Error('用户未登录')
			}

			const response = await request.post('/api/app/address/', {
				...addressData,
				user_id: uid
			})
			
			if (response.data.code === 200) {
				return response.data.data
			} else {
				throw new Error(response.data.message || '添加地址失败')
			}
		} catch (error) {
			console.error('添加地址失败:', error)
			throw error
		}
	}

	/**
	 * 更新地址
	 * @param {number} addressId - 地址ID
	 * @param {Object} addressData - 地址数据
	 * @returns {Promise<Object|null>} 更新结果
	 */
	static async updateAddress(addressId, addressData) {
		try {
			const uid = getUserId()
			if (!uid) {
				throw new Error('用户未登录')
			}

			const response = await request.put(`/api/app/address/${addressId}`, {
				...addressData,
				user_id: uid
			})
			
			if (response.data.code === 200) {
				return response.data.data
			} else {
				throw new Error(response.data.message || '更新地址失败')
			}
		} catch (error) {
			console.error('更新地址失败:', error)
			throw error
		}
	}

	/**
	 * 删除地址
	 * @param {number} addressId - 地址ID
	 * @returns {Promise<boolean>} 删除结果
	 */
	static async deleteAddress(addressId) {
		try {
			const uid = getUserId()
			if (!uid) {
				throw new Error('用户未登录')
			}

			const response = await request.delete(`/api/app/address/${addressId}`)
			
			if (response.data.code === 200) {
				return true
			} else {
				throw new Error(response.data.message || '删除地址失败')
			}
		} catch (error) {
			console.error('删除地址失败:', error)
			throw error
		}
	}

	/**
	 * 设置默认地址
	 * @param {number} addressId - 地址ID
	 * @returns {Promise<boolean>} 设置结果
	 */
	static async setDefaultAddress(addressId) {
		try {
			const uid = getUserId()
			if (!uid) {
				throw new Error('用户未登录')
			}

			const response = await request.put(`/api/app/address/${addressId}/default`, {
				user_id: uid
			})
			
			if (response.data.code === 200) {
				return true
			} else {
				throw new Error(response.data.message || '设置默认地址失败')
			}
		} catch (error) {
			console.error('设置默认地址失败:', error)
			throw error
		}
	}

	/**
	 * 验证地址数据
	 * @param {Object} addressData - 地址数据
	 * @returns {Object} 验证结果 { isValid: boolean, errors: Array }
	 */
	static validateAddress(addressData) {
		const errors = []
		
		if (!addressData.receiver_name || addressData.receiver_name.trim() === '') {
			errors.push('收货人姓名不能为空')
		}
		
		if (!addressData.phone || !/^1[3-9]\d{9}$/.test(addressData.phone)) {
			errors.push('请输入正确的手机号码')
		}
		
		if (!addressData.province || addressData.province.trim() === '') {
			errors.push('省份不能为空')
		}
		
		if (!addressData.city || addressData.city.trim() === '') {
			errors.push('城市不能为空')
		}
		
		if (!addressData.district || addressData.district.trim() === '') {
			errors.push('区县不能为空')
		}
		
		if (!addressData.detail_address || addressData.detail_address.trim() === '') {
			errors.push('详细地址不能为空')
		}
		
		return {
			isValid: errors.length === 0,
			errors
		}
	}
}

export default AddressService